circle.yml

```diff
@@ -379,64 +331,138 @@ jobs:
+        parameters:
+            dockerName:
+                type: string
+                description: Image name to build
+            dockerTag:
+                type: string
+                description: Image tag to build, should match Cypress version, like "3.8.1"
+            workingDirectory:
+                type: string
+                description: Directory with the Dockerfile
+            buildxArches:
+                type: string
+                description: List of architectures to pass to `docker buildx --platform`
+        steps:
+            - checkout
+            - halt-if-docker-image-exists:
+                imageName: << parameters.dockerName >>:<< parameters.dockerTag >>
+            - halt-on-branch
+            - run:
+                ## see https://docs.docker.com/desktop/multi-arch/
+                name: build + push multi-arch image to docker hub
+                working_directory: << parameters.workingDirectory >>
+                command: |
+                    ## install QEMU, pulled from https://github.com/docker/setup-qemu-action/blob/4620c11b701e878e3f2b7aa142927d0646857fd9/src/main.ts
+                    ## avoids https://github.com/docker/buildx/issues/499
+                    BINFMT_IMAGE=tonistiigi/binfmt:latest@sha256:85683def11494bc0055d98dd1081e73cb5e70bbc1ae532be3d12cf054e2b11f1
+                    docker pull $BINFMT_IMAGE
+                    docker image inspect $BINFMT_IMAGE
+                    docker run --rm --privileged $BINFMT_IMAGE --install linux/amd64,linux/arm64
+
+                    ## see https://docs.docker.com/desktop/multi-arch/
+                    echo "$DOCKERHUB_PASS" | docker login -u "$DOCKERHUB_USERNAME" --password-stdin
+                    docker buildx create --name builder --use
+                    docker buildx inspect --bootstrap
+                    DOCKER_TAG=<< parameters.dockerName >>:<< parameters.dockerTag >>
+                    docker buildx build --progress plain --platform << parameters.buildxArches >> -t $DOCKER_TAG --push .
+            - run:
+                name: build + push multi-arch image to amazon ecr
+                working_directory: << parameters.workingDirectory >>
+                command: |
+                    ## now, let's re-build those same images for Amazon ECR
+                    ## see https://docs.aws.amazon.com/AmazonECR/latest/userguide/docker-push-ecr-image.html
+                    aws ecr-public get-login-password --region $AWS_ECR_REGION | docker login --username AWS --password-stdin $AWS_ECR_PREFIX
+                    ## it is inefficient to re-build here - look below for a TODO on how this could be made faster
+                    docker buildx build --progress plain --platform << parameters.buildxArches >> -t $AWS_ECR_PREFIX/$DOCKER_TAG --push .
+
+                    ## TODO: figure out why we can't simply 're-tag' a manifest - the below code gives "authentication required" for some unknown reason
```

Comment:
Do you have any logs from this error? Have you checked that this role has IAM permissions to create a manifest?

Reply:
The logs I have are 1 line that says "authentication required" :stuck_out_tongue: This was with my AWS personal admin account, so it had the required roles. Best I can tell, this is some poorly-documented intersection of the Docker CLI and cross-registry usage - we're basing a Docker image off of a `docker.io` image, but want to push the re-tag to `public.ecr.aws` - since our credentials can never be valid for both, one will say "authentication required". At least that's my guess. I wasn't able to get a clear reason, so I didn't put one here.